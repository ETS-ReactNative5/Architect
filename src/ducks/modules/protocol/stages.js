import uuid from 'uuid/v1';
import { get, compact, set } from 'lodash';
import { arrayMove } from 'react-sortable-hoc';

const CREATE_STAGE = 'PROTOCOL/CREATE_STAGE';
const UPDATE_STAGE = 'PROTOCOL/UPDATE_STAGE';
const UPDATE_PATH = 'PROTOCOL/UPDATE_PATH';
const MOVE_STAGE = 'PROTOCOL/MOVE_STAGE';
const DELETE_STAGE = 'PROTOCOL/DELETE_STAGE';
const DELETE_PROMPT = 'PROTOCOL/DELETE_PROMPT';

const initialState = [];
const initialStage = {
  label: '',
};

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case CREATE_STAGE: {
      const insertAtIndex = get(action, 'index', state.length);

      return [
        ...state.slice(0, insertAtIndex),
        action.stage,
        ...state.slice(insertAtIndex),
      ];
    }
    case UPDATE_STAGE:
      return state.map((stage) => {
        if (stage.id !== action.id) { return stage; }

        if (action.overwrite) { return { ...action.stage, id: stage.id }; }

        return {
          ...stage,
          ...action.stage,
          id: stage.id,
        };
      });
    case UPDATE_PATH: {
      const payload = action.payload;
      return state.map((stage) => {
        if (stage.id !== payload.id) { return stage; }

        const newStage = payload.path ?
          set({ ...stage }, payload.path, payload.value) :
          payload.value;

        return {
          ...newStage,
          id: stage.id,
        };
      });
    }
    case MOVE_STAGE:
      return arrayMove(state, action.oldIndex, action.newIndex);
    case DELETE_STAGE:
      return state.filter(stage => (stage.id !== action.id));
    case DELETE_PROMPT:
      return compact(
        state.map((stage) => {
          if (stage.id !== action.stageId) { return stage; }

          const prompts = stage.prompts.filter(({ id }) => id !== action.promptId);

          // If prompt is empty, we can delete the stage too
          if (action.deleteEmptyStage && prompts.length === 0) { return null; }

          return {
            ...stage,
            prompts,
          };
        }),
      );
    default:
      return state;
  }
}

const createStage = (stage, index) => ({
  type: CREATE_STAGE,
  stage,
  index,
});

const moveStage = (oldIndex, newIndex) => ({
  type: MOVE_STAGE,
  oldIndex,
  newIndex,
});

const updateStage = (stageId, stage, overwrite = false) => ({
  type: UPDATE_STAGE,
  id: stageId,
  stage,
  overwrite,
});

const updatePath = (stageId, path, value) => ({
  type: UPDATE_PATH,
  payload: {
    id: stageId,
    path,
    value,
  },
});

const deleteStage = stageId => ({
  type: DELETE_STAGE,
  id: stageId,
});

const deletePrompt = (stageId, promptId, deleteEmptyStage = false) => ({
  type: DELETE_PROMPT,
  stageId,
  promptId,
  deleteEmptyStage,
});

const createStageThunk = (options, index) =>
  (dispatch) => {
    const stageId = uuid();
    const stage = { ...initialStage, ...options, id: stageId };
    dispatch(createStage(stage, index));
    return stage;
  };

const actionCreators = {
  createStage: createStageThunk,
  updateStage,
  updatePath,
  deleteStage,
  moveStage,
  deletePrompt,
};

const actionTypes = {
  CREATE_STAGE,
  UPDATE_STAGE,
  UPDATE_PATH,
  DELETE_STAGE,
  MOVE_STAGE,
  DELETE_PROMPT,
};

export const test = {
  createStage,
};

export {
  actionCreators,
  actionTypes,
};
