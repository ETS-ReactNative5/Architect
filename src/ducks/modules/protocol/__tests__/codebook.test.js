/* eslint-env jest */

import uuid from 'uuid';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import { sortBy } from 'lodash/fp';
import { actionCreators as formActions } from '../forms';
import { actionCreators as stageActions } from '../stages';
import reducer, { actionCreators, testing } from '../codebook';

jest.mock('uuid');

const sortByType = sortBy('type');

const mockStore = configureStore([thunk]);

const mockState = {
  node: {
    place: { foo: 'bar' },
    person: { hello: 'world' },
  },
  edge: { },
};

describe('protocol.codebook', () => {
  it('does nothing', () => {
    const noop = reducer();
    expect(noop).toEqual({
      node: {},
      edge: {},
    });
  });

  it('createType()', () => {
    const nextState = reducer(
      {
        node: {},
        edge: {},
      },
      actionCreators.createType(
        'node',
        { bazz: 'buzz' },
      ),
    );

    expect(nextState).toEqual({
      node: {
        [uuid()]: { bazz: 'buzz' },
      },
      edge: { },
    });
  });

  it('updateType()', () => {
    const nextState = reducer(
      mockState,
      actionCreators.updateType(
        'node',
        'person',
        { bazz: 'buzz' },
      ),
    );

    expect(nextState).toEqual({
      node: {
        place: { foo: 'bar' },
        person: { bazz: 'buzz' },
      },
      edge: { },
    });
  });

  it('deleteType()', () => {
    const nextState = reducer(
      mockState,
      testing.deleteTypeAction('node', 'person'),
    );

    expect(nextState).toEqual({
      node: {
        place: { foo: 'bar' },
      },
      edge: { },
    });
  });

  it('deleteType() and delete related objects', () => {
    const mockStateWithProtocol = {
      protocol: {
        present: {
          stages: [
            {
              id: 'bazz',
              subject: { entity: 'node', type: 'foo' },
            },
            {
              id: 'buzz',
              prompts: [
                {
                  id: 'fizz',
                  subject: { entity: 'node', type: 'foo' },
                },
              ],
            },
          ],
          forms: {
            bar: {
              entity: 'node',
              type: 'foo',
            },
          },
        },
      },
    };

    const store = mockStore(mockStateWithProtocol);

    store.dispatch(actionCreators.deleteType('node', 'foo', true));

    const actions = store.getActions();

    const expectedActions = sortByType([
      testing.deleteTypeAction('node', 'foo'),
      stageActions.deleteStage('bazz'),
      stageActions.deletePrompt('buzz', 'fizz', true),
      formActions.deleteForm('bar'),
    ]);

    expect(sortByType(actions)).toEqual(expectedActions);
  });
});
