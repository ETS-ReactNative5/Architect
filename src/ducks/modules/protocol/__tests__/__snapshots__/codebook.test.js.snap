// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`protocol.codebook async actions createVariable() dispatches the CREATE_VARIABLE action with a variable id for ego 1`] = `
Object {
  "configuration": Object {
    "fizz": "buzz",
    "name": "bar",
    "type": "text",
  },
  "meta": Object {
    "entity": "ego",
    "type": undefined,
    "variable": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
  },
  "type": "PROTOCOL/CREATE_VARIABLE",
}
`;

exports[`protocol.codebook async actions createVariable() dispatches the CREATE_VARIABLE action with a variable id for node 1`] = `
Object {
  "configuration": Object {
    "fizz": "buzz",
    "name": "bar",
    "type": "text",
  },
  "meta": Object {
    "entity": "node",
    "type": "foo",
    "variable": "809895df-bbd7-4c76-ac58-e6ada2625f9b",
  },
  "type": "PROTOCOL/CREATE_VARIABLE",
}
`;
